core/src/main/java/com/radixdlt/RunningRadixNode.java

```diff
@@ -147,21 +157,36 @@ public void reportSelfStartupTime(Duration startupTimeMs) {
     this.injector.getInstance(Metrics.class).misc().nodeStartup().observe(startupTimeMs);
   }
 
+  public void installShutdownHook() {
+    Thread hook = new Thread(this::shutdownFromHook);
+    final var set = this.shutdownHook.compareAndSet(null, hook);
+    Preconditions.checkState(set, "already installed before");
+    Runtime.getRuntime().addShutdownHook(hook);
+  }
+
   public void shutdown() {
+    final @Nullable Thread hook = this.shutdownHook.getAndSet(null);
+    if (hook != null) {
+      Runtime.getRuntime().removeShutdownHook(hook);
+    }
+    doShutdown("internal reasons");
+  }
+
+  private void shutdownFromHook() {
+    doShutdown("a shutdown hook");
+  }
+
+  private void doShutdown(String reason) {
     // using System.out.printf as logger no longer works reliably in a shutdown hook
-    final var self = injector.getInstance(Key.get(BFTValidatorId.class, Self.class));
-    System.out.printf("Node %s is shutting down...\n", self);
+    System.out.printf("Node %s is shutting down due to %s...\n", this.self(), reason);
```

Comment:
_(note from author)_

This is a drive-by fix: for many tests, this line was throwing Guice errors (since bean of `Key.get(BFTValidatorId.class, Self.class)` is not bound).